pkg/networkmanager/firewall.jsx

```diff
@@ -140,16 +136,41 @@ function ZoneRow(props) {
         deleteButton = <button className="btn btn-danger pficon pficon-delete" onClick={onRemoveZone} />;
     }
 
-    const columns = [
-        { name: props.zone.name, header: true },
-        <>{ props.zone.id === firewall.defaultZone ? <><span className="fa fa-check" /> default</> : '' }</>,
-        <>{ props.zone.interfaces.length > 0 ? props.zone.interfaces.join(', ') : '*' }</>,
-        <>{ props.zone.source.length > 0 ? props.zone.source.join(', ') : '*' }</>,
-        deleteButton,
-    ];
-    return <ListingRow key={props.zone.id}
-                       rowId={props.zone.id}
-                       columns={columns} />;
+    var addServiceAction;
+    if (firewall.readonly) {
+        addServiceAction = (
+            <OverlayTrigger placement="top"
+                                overlay={ <Tooltip id="tip-auth">{ _("You are not authorized to modify the firewall.") }</Tooltip> }>
+                <Button bsStyle="primary" className="pull-right" disabled> {_("Add Services")} </Button>
+            </OverlayTrigger>
+        );
+    } else {
+        addServiceAction = (
+            <Button bsStyle="primary" onClick={() => props.openServicesDialog(props.zone.id, props.zone.name)} id="add-services-button">
+                {_("Add Services")}
+            </Button>
+        );
+    }
+
+    return <div className="zone-section" data-id={props.zone.id}>
+        <h4>
+            <span>
+                <strong>{ cockpit.format(_("$0 Zone"), props.zone.name) }</strong>
+                { props.zone.interfaces.length > 0 && <span>&nbsp;&nbsp;&nbsp;<strong>{_("Interfaces")}</strong> {props.zone.interfaces.join(", ")}</span> }
+                { props.zone.source.length > 0 && <span>&nbsp;&nbsp;&nbsp;<strong>{_("Addresses")}</strong> 111.111.111.111/24{props.zone.source.join(", ")}</span> }
+            </span>
+            <span className="btn-group">{deleteButton}&nbsp;{addServiceAction}</span>
```

Comment:
Using btn-group here will make them look too related to each other. Needs some spacing between them

Reply:
Fixed in https://github.com/andreasn/cockpit/commit/1bfc0a9c98f1b396f19b8553ded42a17386d0f32

![Screenshot from 2019-09-27 11-56-58](https://user-images.githubusercontent.com/209895/65760957-3e2e4900-e11e-11e9-8904-e899e71682b6.png)

The implementation is a bit dirty though. Now it will give this property only to a delete button inside that span. Would probably be better to create a class with the margin and then apply that to that button specifically instead strictly to the btn-danger element. Or maybe to first-child in that span? Not sure what's best.

Reply:
@garrett any tips on how to solve this in a clean way?